cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/lineage/LineageOperationsProcessor.java

```diff
@@ -129,34 +150,70 @@ public LineageOperationsProcessor(Set<Connection> stageConnections,
+   */
+  private void addMergeOperation(Set<String> stageInputs, Map<String, Operation> processedOperations) {
     Set<String> sortedInputs = new TreeSet<>(stageInputs);
-    String mergeOperationName = prefixedOperationName(Joiner.on(SEPARATOR).join(sortedInputs), "merge");
+    String mergeOperationNamePrefix = getMergeOperationNamePrefix(sortedInputs);
     String mergeDescription = "Merged stages: " + Joiner.on(",").join(sortedInputs);
-    if (processedOperations.containsKey(mergeOperationName)) {
-      // it is possible that same stages act as an input to multiple stages.
-      // we should still only add single merge operation for them
-      return;
-    }
-    List<InputField> inputFields = new ArrayList<>();
+
+    Map<String, List<InputField>> fieldNameMap = new LinkedHashMap<>();
+    // create the map that contains the field name of the current stages and the parent stages to the origins
     for (String inputStage : sortedInputs) {
       List<String> parentStages = findParentStages(inputStage);
+      // traverse in a reverse order of parent stages since the parent stages will contain the closest parent
+      // at the end of the list
+      Collections.reverse(parentStages);
+      // this stores the visited field, if we already know the field from the previous parent, do not
+      // add other origin of this field
+      Set<String> visitedField = new HashSet<>();
       for (String parentStage : parentStages) {
+        // get the map of all the outputs to the origin map from a stage
         Map<String, String> fieldOrigins = stageOutputsWithOrigins.get(parentStage);
         for (Map.Entry<String, String> fieldOrigin : fieldOrigins.entrySet()) {
-          inputFields.add(InputField.of(fieldOrigin.getValue(), fieldOrigin.getKey()));
+          String fieldName = fieldOrigin.getKey();
+          if (visitedField.contains(fieldName)) {
+            continue;
+          }
+          List<InputField> inputFields = fieldNameMap.computeIfAbsent(fieldName, k -> new ArrayList<>());
+          inputFields.add(InputField.of(fieldOrigin.getValue(), fieldName));
+          visitedField.add(fieldName);
         }
       }
     }
 
-    Set<String> outputs = new LinkedHashSet<>();
-    for (InputField inputField : inputFields) {
-      outputs.add(inputField.getName());
-    }
-    TransformOperation merge = new TransformOperation(mergeOperationName, mergeDescription, inputFields,
-                                                      new ArrayList<>(outputs));
-    processedOperations.put(merge.getName(), merge);
+    fieldNameMap.forEach((fieldName, inputFields) -> {
+      String mergeName = prefixedName(mergeOperationNamePrefix, fieldName);
+      if (processedOperations.containsKey(mergeName)) {
+        // it is possible that same stages act as an input to multiple stages.
+        // we should still only add single merge operation for them
+        return;
+      }
+
+      Set<String> outputs = new LinkedHashSet<>();
```

Comment:
this makes it look like it's still a many to many operation. Would be better to change the `Map<String, List<InputField>>` into a Map<String, Set<String>>` so that it is clear there is only ever one output, which is the fieldName.

Reply:
Changed this to a ` Map<String, List<String>>` , where the value is list of origins of the field, a list better matches the API since the transform operation expects a list instead of a set